presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDriver.java

```diff
@@ -57,31 +79,332 @@
+            }
+
+            private void releaseForCleaning()
+            {
+                --cleanerCount;
+            }
+        }
+
+        private final ScheduledExecutorService valueCleanupService;
+        private final LoadingCache<K, Holder> backingCache;
+        private final AtomicBoolean closed = new AtomicBoolean(false);
+
+        private ReferenceCountingLoadingCache(CacheLoader<K, V> loader, Consumer<V> disposer)
+        {
+            this.valueCleanupService = new ScheduledThreadPoolExecutor(1, daemonThreadsNamed("cache-cleanup-%s"));
+            this.backingCache = CacheBuilder.newBuilder()
+                    .removalListener(new RemovalListener<K, Holder>() {
+                        @Override
+                        public void onRemoval(RemovalNotification<K, Holder> notification)
+                        {
+                            Holder holder = notification.getValue();
+                            /*
+                             * The docs say that both the key and value may be
+                             * null if they've already been garbage collected.
+                             * We aren't using weak or soft keys or values, so
+                             * this shouldn't apply.
+                             */
+                            requireNonNull(holder, format("holder is null while removing key %s", notification.getKey()));
+                            checkState(holder.getRefcount() == 0, "Non-zero refcount disposing %s", notification.getKey());
+                            checkState(holder.getCleanerCount() == 0, "Non-zero refcount disposing %s", notification.getKey());
+                            disposer.accept(holder.get());
+                        }
+                    })
+                    .build(new CacheLoader<K, Holder>() {
+                        @Override
+                        public Holder load(K key)
+                                throws Exception
+                        {
+                            return new Holder(loader.load(key));
+                        }
+                    });
+        }
+
+        private void close()
+        {
+            if (closed.compareAndSet(false, true)) {
+                valueCleanupService.shutdownNow();
+                backingCache.invalidateAll();
+            }
+        }
+
+        private V acquire(K key)
+        {
+            synchronized (this) {
+                Holder holder = backingCache.getUnchecked(key);
+                holder.reference();
+                return holder.get();
+            }
+        }
+
+        private void release(K key)
+        {
+            /*
+             * Access through the Map interface to avoid creating a Holder and immediately
+             * scheduling it for cleanup.
+             */
+            Holder holder = backingCache.asMap().get(key);
+            if (holder == null) {
+                return;
+            }
+
+            Runnable deferredRelease = () -> {
```

Comment:
I could be talked into moving this into `Holder.scheduleCleanup()`. That would require passing in `key` since Holder doesn't have it already.